tools/jnimarshalmethod-gen/App.cs

```diff
@@ -56,43 +114,96 @@ static void CreateMarshalMethodAssembly (string path)
 			foreach (var type in assembly.DefinedTypes) {
+				if (type.IsGenericType || type.IsGenericTypeDefinition)
+					continue;
+
+				var td                      = ad.MainModule.FindType (type);
+
+				if (td == null) {
+					if (verbose)
+						Warning ($"Unable to find cecil's TypeDefinition of type {type}");
+					continue;
+				}
+
 				var registrationElements    = new List<Expression> ();
 				var targetType              = Expression.Variable (typeof(Type), "targetType");
 				TypeBuilder dt = null;
 
 				var flags = BindingFlags.Public | BindingFlags.NonPublic |
 						BindingFlags.Instance | BindingFlags.Static;
-				foreach (var method in type.GetMethods (flags )) {
-					// TODO: Constructors, [Register] methods
+				foreach (var method in type.GetMethods (flags)) {
+					// TODO: Constructors
 					var export  = method.GetCustomAttribute<JavaCallableAttribute> ();
-					if (export == null)
-						continue;
+					string signature = null;
+					string name = null;
+					string methodName = method.Name;
+
+					if (export == null) {
+						if (method.IsGenericMethod || method.ContainsGenericParameters || method.IsGenericMethodDefinition || method.ReturnType.IsGenericType)
+							continue;
+
+						if (method.DeclaringType != type || !td.ImplementsInterface ("Java.Interop.IJavaPeerable"))
+							continue;
+
+						var md = td.GetMethodDefinition (method);
+
+						if (md == null) {
+							if (verbose)
+								Warning ($"Unable to find cecil's MethodDefinition of method {method}");
+							continue;
+						}
+
+						if (!md.NeedsMarshalMethod (resolver, method, ref name, ref methodName, ref signature))
+							continue;
+					}
+
 					if (dt == null)
-						dt = dm.DefineType (type.FullName, TypeAttributes.Public | TypeAttributes.Sealed);
+						dt = GetTypeBuilder (dm, type);
+
+					if (verbose)
+						Console.WriteLine ($"Adding marshal method for {Terminal.FgColor (Color.Green)}{method}{Terminal.ResetColor ()}");
 
 					var mb = dt.DefineMethod (
-							method.Name,
-							MethodAttributes.Public | MethodAttributes.Static);
+							methodName,
+							System.Reflection.MethodAttributes.Public | System.Reflection.MethodAttributes.Static);
+
 					var lambda  = builder.CreateMarshalToManagedExpression (method);
 					lambda.CompileToMethod (mb);
-					var signature = export.Signature ??
-							builder.GetJniMethodSignature (method);
-					registrationElements.Add (CreateRegistration (export.Name, signature, lambda, targetType, method.Name));
+
+					if (export != null) {
+						name = export.Name;
+						signature = export.Signature;
+					}
+
+					if (signature == null)
+						signature = builder.GetJniMethodSignature (method);
+
+					registrationElements.Add (CreateRegistration (name, signature, lambda, targetType, methodName));
 				}
-				if (dt != null) {
+				if (dt != null)
 					AddRegisterNativeMembers (dt, targetType, registrationElements);
-					dt.CreateType ();
-				}
 			}
+
+			foreach (var tb in definedTypes)
```

Comment:
I wonder if there's an ordering relation here, e.g. do nested types need to be created before their enclosing types, or vice-versa?

Reply:
The order is not important here as everything is already prepared at that point. It is important at the time we add nested types and that is why we postpone `CreateType`.